### PR TITLE
pre_defined_classes-simple_programs_taking_string_user_input-medium-q7

### PR DESCRIPTION
**Pull request by yuqi88**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
---

## Not yet, but soon

- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/UrlValidator.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package pre_defined_classes.simple_programs_taking_string_user_input.medium.q7;
+
+/*
+A website address typically begins with <tt>www.</tt> and ends in <tt>.ca</tt>, <tt>.com</tt>, <tt>.edu</tt>,
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/UrlValidator.java:4]: You can remove the Html tags

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/UrlValidator.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+among others. Originally, the ending of the address signifies the type of website it is.
+For example, the ending <tt>.ca</tt> is used to designate Canadian websites (although Canadian websites do not have
+to use this ending at all). Make a program that answers <tt>true</tt> or <tt>false</tt> to the question of whether
+the inputted address is a Canadian website.
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/UrlValidator.java:8]: Should add `Assume the user will enter a valid URL.`

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        return Stream.of("https://www.canada.ca" , "canvas.ubc.ca", "ubc.ca", "made.up.ca");
+    }
+
+    public static Stream<String> noneCanadianWebsite() {
+        return Stream.of("https://www.gouvernement.fr", "just.a.string", "coca-cola.com", "testing.ca.com");
+    }
+
+    @ParameterizedTest
+    @MethodSource("canadianWebsite")
+    public void correctlyDetectCanadaWebsites(String caUrl) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:40]: *Detects

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly print true when a url ends with .ca";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:41]: I would say `Your program does not print true when the URL is a Canadian website. `

Comment (Hedgemon4) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:41]: We want the error messages to be helpful, but not give away the exact answer

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        runWithInput(caUrl, new Clause[]{
+                new StringLiteral("true"),
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("noneCanadianWebsite")
+    public void correctlyDetectNoneCanadaWebsites(String notCaUrl) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly print false when a url does not end with .ca";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:50]: Same as other error message

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java (hunk 4 above)
Comment (yuqi88) [reply src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:41]: okie

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public static Stream<String> noneCanadianWebsite() {
+        return Stream.of("https://www.gouvernement.fr", "just.a.string", "coca-cola.com", "testing.ca.com");
+    }
+
+    @ParameterizedTest
+    @MethodSource("canadianWebsite")
+    public void correctlyDetectsCanadaWebsites(String caUrl) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print true when the URL is a Canadian website.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:41]: I know I said to do it this way, but I think the print true gives too much feedback, so it should maybe be `does not print the correct output when`

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        runWithInput(caUrl, new Clause[]{
+                new StringLiteral("true"),
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("noneCanadianWebsite")
+    public void correctlyDetectsNoneCanadaWebsites(String notCaUrl) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print false when the URL is not a Canadian website.";
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:50]: Same as the other error message

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
Other than the small error message change, this one looks good

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        return Stream.of("https://www.canada.ca" , "canvas.ubc.ca", "ubc.ca", "made.up.ca");
+    }
+
+    public static Stream<String> noneCanadianWebsite() {
+        return Stream.of("https://www.gouvernement.fr", "just.a.string", "coca-cola.com", "testing.ca.com");
+    }
+
+    @ParameterizedTest
+    @MethodSource("canadianWebsite")
+    public void correctlyDetectsCanadaWebsites(String caUrl) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:40]: Method name should be `correctlyDetectsCanadianWebsites` to be grammatically correct

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the correct output when the URL is a Canadian website.";
+        runWithInput(caUrl, new Clause[]{
+                new StringLiteral("true"),
+        });
+    }
+
+    @ParameterizedTest
+    @MethodSource("noneCanadianWebsite")
+    public void correctlyDetectsNoneCanadaWebsites(String notCaUrl) throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:49]: `correctlyDetectsNonCanadianWebsites`

File: src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public void runMain(){ UrlValidator.main(new String[0]);}
+
+    public static Stream<String> canadianWebsite() {
+        return Stream.of("https://www.canada.ca" , "canvas.ubc.ca", "ubc.ca", "made.up.ca");
+    }
+
+    public static Stream<String> noneCanadianWebsite() {
+        return Stream.of("https://www.gouvernement.fr", "just.a.string", "coca-cola.com", "testing.ca.com");
```
Comment (Hedgemon4) [inline src/pre_defined_classes/simple_programs_taking_string_user_input/medium/q7/MainTest.java:35]: Add System.lineSperator()

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**